File: Tests/EscortRetrievingIntervalsSpec.m
#import "Kiwi.h"
#import "FakeDateUtil.h"
#import "NSDate+AZDateBuilder.h"
#import "NSDate+Escort.h"

SPEC_BEGIN(EscortRetrievingIntervals)
    registerMatchers(@"AZ");// NSDate Custom Matcher

    describe(@"-minutesAfterDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 10 minutes ago", ^{
                int tenMinutes = 10;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingMinutes:tenMinutes];
                });
                it(@"should return 10", ^{
                    NSInteger minutes = [currentDate minutesAfterDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(tenMinutes)];
                });
            });
            context(@"when 1 hour ago", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingHours:oneHour];
                });
                it(@"should return 60", ^{
                    NSInteger minutes = [currentDate minutesAfterDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(60)];
                });
            });
            context(@"when 10 minutes later", ^{
                int tenMinutes = 10;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingMinutes:tenMinutes];
                });
                it(@"should return -10", ^{
                    NSInteger minutes = [currentDate minutesAfterDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(-tenMinutes)];
                });
            });
        });
    });

    describe(@"-minutesBeforeDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 10 minutes later", ^{
                int tenMinutes = 10;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingMinutes:tenMinutes];
                });
                it(@"should return 10", ^{
                    NSInteger minutes = [currentDate minutesBeforeDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(tenMinutes)];
                });
            });
            context(@"when 1 hour later", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingHours:oneHour];
                });
                it(@"should return 60", ^{
                    NSInteger minutes = [currentDate minutesBeforeDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(60)];
                });
            });
            context(@"when 10 minutes ago", ^{
                int tenMinutes = 10;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingMinutes:tenMinutes];
                });
                it(@"should return -10", ^{
                    NSInteger minutes = [currentDate minutesBeforeDate:theOtherDate];
                    [[theValue(minutes) should] equal:theValue(-tenMinutes)];
                });
            });
        });
    });

    describe(@"-hoursAfterDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 1 hour ago", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingHours:oneHour];
                });
                it(@"should return 1", ^{
                    NSInteger result = [currentDate hoursAfterDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(oneHour)];
                });
            });
            context(@"when 1 day ago", ^{
                int oneDay = 24;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingHours:oneDay];
                });
                it(@"should return 24", ^{
                    NSInteger result = [currentDate hoursAfterDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(oneDay)];
                });
            });
            context(@"when 1 hour later", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingHours:oneHour];
                });
                it(@"should return -1", ^{
                    NSInteger result = [currentDate hoursAfterDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(-oneHour)];
                });
            });
        });
    });

    describe(@"-hoursBeforeDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 1 hour later", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingHours:oneHour];
                });
                it(@"should return 1", ^{
                    NSInteger result = [currentDate hoursBeforeDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(oneHour)];
                });
            });
            context(@"when 1 day later", ^{
                int oneDay = 24;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingHours:oneDay];
                });
                it(@"should return 24", ^{
                    NSInteger result = [currentDate hoursBeforeDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(oneDay)];
                });
            });
            context(@"when 1 hour ago", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingHours:oneHour];
                });
                it(@"should return -1", ^{
                    NSInteger result = [currentDate hoursBeforeDate:theOtherDate];
                    [[theValue(result) should] equal:theValue(-oneHour)];
                });
            });
        });
    });


    describe(@"-daysAfterDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 1 day ago", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingDays:oneHour];
                });
                it(@"should return 1", ^{
                    NSInteger day = [currentDate daysAfterDate:theOtherDate];
                    [[theValue(day) should] equal:theValue(oneHour)];
                });
            });
            context(@"when 1 day later", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingDays:oneHour];
                });
                it(@"should return -1", ^{
                    NSInteger day = [currentDate daysAfterDate:theOtherDate];
                    [[theValue(day) should] equal:theValue(-oneHour)];
                });
            });
        });
    });

    describe(@"-daysBeforeDate", ^{
        context(@"the date is 2010-10-10 10:10:10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @10,
                    AZ_DateUnit.second : @10,
                }];
                [FakeDateUtil stubCurrentDate:currentDate];
            });
            context(@"when 1 day ago", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateBySubtractingDays:oneHour];
                });
                it(@"should return -1", ^{
                    NSInteger day = [currentDate daysBeforeDate:theOtherDate];
                    [[theValue(day) should] equal:theValue(-oneHour)];
                });
            });
            context(@"when 1 day later", ^{
                int oneHour = 1;
                __block NSDate *theOtherDate;
                beforeEach(^{
                    theOtherDate = [currentDate dateByAddingDays:oneHour];
                });
                it(@"should return 1", ^{
                    NSInteger day = [currentDate daysBeforeDate:theOtherDate];
                    [[theValue(day) should] equal:theValue(oneHour)];
                });
            });
        });
    });
    SPEC_END

File: SwiftTest/EscortRetrievingIntervalsSpec.swift
import XCTest
import Quick
import AZDateBuilder

class EscortRetrievingIntervalsSpec: QuickSpec {
    override func spec() {
        describe("seconds(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 seconds should return 5") {
                let anotherDate = currentDate.add(second: 5)
                XCTAssertEqual(currentDate.seconds(after: anotherDate), 5)
            }
            it("after 10 minutes should return 600") {
                let anotherDate = currentDate.add(minute: 10)
                XCTAssertEqual(currentDate.seconds(after: anotherDate), 600)
            }
        }
        describe("minute(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 miuntes should return 5") {
                let anotherDate = currentDate.add(minute: 5)
                XCTAssertEqual(currentDate.minutes(after: anotherDate), 5)
            }
            it("after 10 hours should return 600") {
                let anotherDate = currentDate.add(hour: 10)
                XCTAssertEqual(currentDate.minutes(after: anotherDate), 600)
            }
        }
        describe("hours(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 hours should return 5") {
                let anotherDate = currentDate.add(hour: 5)
                XCTAssertEqual(currentDate.hours(after: anotherDate), 5)
            }
            it("after 1 days should return 24") {
                let anotherDate = currentDate.add(day: 1)
                XCTAssertEqual(currentDate.hours(after: anotherDate), 24)
            }
        }
        describe("days(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 days should return 5") {
                let anotherDate = currentDate.add(day: 5)
                XCTAssertEqual(currentDate.days(after: anotherDate), 5)
            }
            it("after 2 months should return 61") {
                let anotherDate = currentDate.add(month: 2)
                XCTAssertEqual(currentDate.days(after: anotherDate), 61)
            }
        }
        describe("months(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 months should return 5") {
                let anotherDate = currentDate.add(month: 5)
                XCTAssertEqual(currentDate.months(after: anotherDate), 5)
            }
            it("after 2 years should return 24") {
                let anotherDate = currentDate.add(year: 2)
                XCTAssertEqual(currentDate.months(after: anotherDate), 24)
            }
        }
        describe("years(after:)") {
            let currentDate = Date.date(by: [
                .year: 2010,
                .month: 10,
                .day: 11,
                .hour: 12,
                .minute: 13,
                .second: 14,
                ])
            it("after 5 years should return 5") {
                let anotherDate = currentDate.add(year: 5)
                XCTAssertEqual(currentDate.years(after: anotherDate), 5)
            }
        }
    }
}
